openhtf/__init__.py

```diff
@@ -45,6 +46,105 @@ class InvalidTestPhaseError(Exception):
+        # Have to wrap with lambda: or else the trigger gets called in __init__.
+        'test_start': lambda: triggers.AutoStart,
+        'test_stop': lambda: triggers.AutoStop,
+        'output_callbacks': list,
+        # pylint: disable=used-before-assignment
+        'code_info': lambda: CodeInfo.ForModuleFromStack(levels_up=5),
+    })):
+  """An object that represents an OpenHTF test.
+
+  This object encapsulates the static test information that is set once and used
+  by the framework along the way.
+
+  Example:
+
+    def PhaseOne(test):
+      # Integrate more widgets
+
+    def PhaseTwo(test):
+      # Analyze widget integration status
+
+    Test(PhaseOne, PhaseTwo).Execute()
+
+    # Or more directly:
+    TestInfo([PhaseOne, PhaseTwo]).Execute()
+
+  Attributes:
+    phases: The phases to execute for this test.
+    metadata: Any metadata that should be associated with test records.
+    code_info: Information about the module that created the test.
+    test_start: Trigger for starting the test, defaults to AutoStart with a
+        dummy serial number.
+    test_stop: Trigger for when the test is over, defaults to AutoStop to
+        immediately stop after the phase.
+    output_callbacks: List of callbacks to be called with the results
+        output from this test.
+  """
+
+  def __init__(self, *args, **kwargs):
+    super(TestInfo, self).__init__(*args, **kwargs)
+    self.phases = [TestPhaseInfo.WrapOrReturn(phase) for phase in self.phases]
+
+  def AddOutputCallback(self, callback):
+    """Add the given function as an output module to this test."""
+    self.output_callbacks.append(callback)
+
+  def OutputTestRecord(self, test_record):
+    """Feed the record of this test to all output modules."""
+    test_record.metadata.update(self.metadata)
+    for output_cb in self.output_callbacks:
+      output_cb(test_record)
+
+  def GetPlugTypeMap(self):
+    """Returns dict mapping name to plug type for all phases."""
+    plug_type_map = {}
+    for plug, plug_type in itertools.chain.from_iterable(
+        ((plug.name, plug.cls) for plug in phase.plugs)
+        for phase in self.phases):
+      if (plug in plug_type_map and
+          plug_type is not plug_type_map[plug]):
+        raise plugs.DuplicatePlugError(
+            'Duplicate plug with different type: %s' % plug)
+      plug_type_map[plug] = plug_type
+    return plug_type_map
+
+  def Execute(self):
+    """Starts the framework and executes the given test."""
+    SetupFramework()
+
+    _LOG.info('Executing test: %s', self.code_info.name)
+    executor = exe.TestExecutor(
+        conf.Config(), self, self.test_start, self.test_stop)
+    server = http_api.Server(executor)
+    StopOnSigInt([server.Stop, executor.Stop])
+    server.Start()
+    executor.Start()
+    executor.Wait()
+    server.Stop()
+
+
+def SetupFramework():
+  """Sets up various bits of the framework. Only needs to be called once."""
+  try:
+    FLAGS(sys.argv)  # parse flags
```

Comment:
ah, yeah it was wrong before, just nobody noticed because it's uncommon to use other argv args, but you can see from a gflags docstring here:
https://github.com/google/python-gflags/blob/master/gflags/flagvalues.py#L75